### ISSUE TITLE
Tshark, ndpi.lua and Application Latency

### ISSUE THREAD

**Issue by alisoho** [ntop/nDPI#978]
Hi,

I am trying to use ndpi.lua scirpt in Tshark for Application Latency. I have test it with Wireshark and it works perfectly (Thanks!) but it doesn't work with Tshark. 
As for first step, I tried to modify "appl_rtt_dialog_menu()" function in order to print the Application latency in the console, simply like this:
```
local function appl_rtt_dialog_menu()
    local label = ""
    local i
    i = 0
    label = label .. "Server\t\tMin Application RTT\n"
    for k,v in pairsByValues(min_appl_RRT, rev) do
        print("test")
        label = label .. string.format("%-20s\t%.3f / %.3f msec\n", shortenString(k), v, max_appl_RRT[k])
        if(i == max_num_entries) then break else i = i + 1 end
    end
    io.write(label)
end
appl_rtt_dialog()
```
but I receive no output except "Server" and "Application RTT" tag names.
I tracked down the problem and I realized, in Tshark , when it executes `pairsByValues()` function, in iterator function it returns "nil" and actually `a[i] == nil` but in Wireshark it doesn't. 
```
function pairsByValues(t, f)
    local a = {}
    for n in pairs(t) do table.insert(a, n) end
    table.sort(a, function(x, y) return f(t[x], t[y]) end)
    local i = 0      -- iterator variable
    local iter = function ()   -- iterator function
        i = i + 1
        if a[i] == nil then return nil
        else return a[i], t[a[i]]
        end
    end
    return iter
end
```
what is the causes of this problem?
is there any inconsistency between Wireshark and Tshark for loading lua plugins?
How can I fix it?

**Comment by alisoho** [comment]
I figured out that lua scripts run before packets are read in Wireshark/Tshark as mentioned here: [https://wiki.wireshark.org/Lua](url)
so the ndpi.lua is useless with Tshark.

**Comment by IvanNardi** [comment]
Hi,

I have never used ndpi.lua with tshark, I don't quite understand what you want to achieve but I'd like to share some thoughts 

There are some issues while using ndpi.lua plugin with tshark, but, at the end of the day, it works (more or less).
AFAIK, the main source of misunderstanding is the two-pass analysis:
* this lua script ignores the first-pass and processes the packets only during the second one;
* the "ndpi" virtual extcap interface is a "live" one and it performs a live capture even when opening a pcap file.

With wireshark none of this stuff is relevant, but tshark doesn't support two-pass analysis on live captures, so ndpi.lua script is (almost) never triggered.

As a workaround, you can force the ndpi.lua script to elaborate the packet on the first pass: change "if(pinfo.visited == true)" in "if(pinfo.visited == false)" around line 900.
I have no idea what ndpi information you lose this way, but you should get most of them, at least the basic ones about protocol classification.
Example:

./tshark -i ndpi -o extcap.ndpi.i:/home/ivan/pcaps/ssh.pcap -o extcap.ndpi.9:-1 -Y "ndpi.protocol.name == SSH"

Bottom line: you need to tweak the script to be sure to calculate the information you want on the first pass

Hope that this help

**Comment by alisoho** [comment]
> Hi,
> 
> I have never used ndpi.lua with tshark, I don't quite understand what you want to achieve but I'd like to share some thoughts
> 
> There are some issues while using ndpi.lua plugin with tshark, but, at the end of the day, it works (more or less).
> AFAIK, the main source of misunderstanding is the two-pass analysis:
> 
> * this lua script ignores the first-pass and processes the packets only during the second one;
> * the "ndpi" virtual extcap interface is a "live" one and it performs a live capture even when opening a pcap file.
> 
> With wireshark none of this stuff is relevant, but tshark doesn't support two-pass analysis on live captures, so ndpi.lua script is (almost) never triggered.
> 
> As a workaround, you can force the ndpi.lua script to elaborate the packet on the first pass: change "if(pinfo.visited == true)" in "if(pinfo.visited == false)" around line 900.
> I have no idea what ndpi information you lose this way, but you should get most of them, at least the basic ones about protocol classification.
> Example:
> 
> ./tshark -i ndpi -o extcap.ndpi.i:/home/ivan/pcaps/ssh.pcap -o extcap.ndpi.9:-1 -Y "ndpi.protocol.name == SSH"
> 
> Bottom line: you need to tweak the script to be sure to calculate the information you want on the first pass
> 
> Hope that this help

Hi,

thanks for the detailed information.
my target is to get the Apllication Latancy with Tshark +ndpi.lua 
I am not trying to use ndpi for filtering any specific protocol.  Simply, I just want this Wireshark output in Tshark:
![image](https://user-images.githubusercontent.com/69587578/90126415-eadbde00-dd63-11ea-8f23-3f2e00eea075.png)
the major problem is lua scripts run before packets are read in Wireshark/Tshark so in Wireshark it is possible to invoke functions in the script through GUI but in Tshark it is not possible.

**Comment by IvanNardi** [comment]
> the major problem is lua scripts run before packets are read in Wireshark/Tshark so in Wireshark it is possible to invoke functions in the script through GUI but in Tshark it is not possible.

Hi,
Fortunately, you are wrong because it is definitely possible to execute (part of) ndpi.lua from tshark.
Try changing your ndpi.lua file according to the attached diff file: it is nothing more than a quick and dirty hack, but it should give you what you want

[ndpi.lua.diff.txt](https://github.com/ntop/nDPI/files/5071035/ndpi.lua.diff.txt)

**Comment by alisoho** [comment]
Perfect!
thanks, now it is working.

**Comment by DD152** [comment]
> Perfect!
> thanks, now it is working.

Hi,
Gald to see that you hava solved the problem.It means that I maybe find a way to clarify my doubt.
I also want to get Aplication\Network Latancy.So I want to figure out  how to get Aplication\Network Latancy with the ndpi.lua&tshark in terminal?
Please,thanks.
